File: ReactiveCollection/Sources/TextField/UITextField+RAC.m
//
//  UITextField+RAC.m
//  ReactiveCollection
//
//  Created by Paweł Sękara on 31.08.2016.
//  Copyright © 2016 Codewise Sp. z o.o. Sp. K. All rights reserved.
//

#import "UITextField+RAC.h"
#import <objc/runtime.h>
#import <ReactiveUIBindings/ReactiveUIBindings-Swift.h>

@implementation UITextField (RAC)

+ (void)load {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        Class clazz = [self class];
        
        Method originalMethod = class_getInstanceMethod(clazz, @selector(setDelegate:));
        Method swizzledMethod = class_getInstanceMethod(clazz, @selector(rac_setDelegate:));
        
        method_exchangeImplementations(originalMethod, swizzledMethod);
    });
}

- (void)rac_setDelegate:(id<UITextFieldDelegate>)delegate {
    if (self.delegate != nil) {
        if ([self.delegate isKindOfClass:TextFieldDelegateProxy.class] && (![delegate isKindOfClass:TextFieldDelegateProxy.class])) {
            ((TextFieldDelegateProxy *) self.delegate).forwardDelegate = delegate;
        } else if (![self.delegate isKindOfClass:TextFieldDelegateProxy.class] && ([delegate isKindOfClass:TextFieldDelegateProxy.class])) {
            ((TextFieldDelegateProxy *) delegate).forwardDelegate = self.delegate;
            [self rac_setDelegate:delegate];
        }
    } else {
        [self rac_setDelegate:delegate];
    }
}

@end

File: ReactiveCollection/Sources/TextField/UITextField+RAC.swift
//
//  UITextField+RAC.swift
//  ReactiveCollection
//
//  Created by Paweł Sękara on 30.08.2016.
//  Copyright © 2016 Codewise Sp. z o.o. Sp. K. All rights reserved.
//

import Foundation
import UIKit
import ReactiveSwift
import Result


public extension UITextField {
    
    public weak var forwardDelegate: UITextFieldDelegate? {
        get {
            guard let proxy = self.delegate as? TextFieldDelegateProxy else {
                return nil
            }
            return proxy.forwardDelegate as? UITextFieldDelegate
        }
        set {
            let proxy = TextFieldDelegateProxy.proxy(forObject: self)
            proxy.forwardDelegate = newValue as? NSObject
        }
    }
        
}

public extension Reactive where Base: UITextField {
    
    public var editingStarted: Signal<Void, NoError> {
        let proxy = TextFieldDelegateProxy.proxy(forObject: self.base)
        return proxy.rac_editingStarted
    }
    
    public var editingEnded: Signal<Void, NoError> {
        let proxy = TextFieldDelegateProxy.proxy(forObject: self.base)
        return proxy.rac_editingEnded
    }
    
    public var textSignal: SignalProducer<String, NoError> {
        let proxy = TextFieldDelegateProxy.proxy(forObject: self.base)
        return proxy.rac_textSignal
    }
}
